third_party/tzcode/strftime.c

```diff
@@ -0,0 +1,679 @@
+	const char *wday[DAYSPERWEEK];
+	const char *weekday[DAYSPERWEEK];
+	const char *X_fmt;
+	const char *x_fmt;
+	const char *c_fmt;
+	const char *am;
+	const char *pm;
+	const char *date_fmt;
+};
+
+#define Locale (&C_time_locale)
+
+static const struct lc_time_T C_time_locale = {
+	{"Jan", "Feb", "Mar", "Apr", "May", "Jun",
+	 "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"},
+	{"January", "February", "March", "April", "May", "June",
+	 "July", "August", "September", "October", "November", "December"},
+	{"Sun", "Mon", "Tue", "Wed",
+	 "Thu", "Fri", "Sat"},
+	{"Sunday", "Monday", "Tuesday", "Wednesday",
+	 "Thursday", "Friday", "Saturday"},
+
+	/* X_fmt */
+	"%H:%M:%S",
+
+	/*
+	 ** x_fmt
+	 ** C99 and later require this format.
+	 ** Using just numbers (as here) makes Quakers happier;
+	 ** it's also compatible with SVR4.
+	 */
+	"%m/%d/%y",
+
+	/*
+	 ** c_fmt
+	 ** C99 and later require this format.
+	 ** Previously this code used "%D %X", but we now conform to C99.
+	 ** Note that
+	 ** "%a %b %d %H:%M:%S %Y"
+	 ** is used by Solaris 2.3.
+	 */
+	"%a %b %e %T %Y",
+
+	/* am */
+	"AM",
+
+	/* pm */
+	"PM",
+
+	/* date_fmt */
+	"%a %b %e %H:%M:%S %Z %Y"
+};
+
+static int pow10[] = { 1,      10,	100,	  1000,	     10000,
+		       100000, 1000000, 10000000, 100000000, 1000000000 };
+
+enum warn {
+	IN_NONE,
+	IN_SOME,
+	IN_THIS,
+	IN_ALL
+};
+
+static char *
+_add(const char *, char *, const char *);
+static char *
+_conv(int, const char *, char *, const char *);
+static char *
+_fmt(const char *, const struct datetime_tm *, char *, const char *, enum warn *);
+static char *
+_yconv(int, int, bool, bool, char *, char const *);
+
+/*
+ * Given the seconds from Epoch (1970-01-01) we calculate date
+ * since Rata Die (0001-01-01).
+ * DT_EPOCH_1970_OFFSET is the distance in days from Rata Die to Epoch.
+ */
+static int
+local_dt(int64_t secs)
+{
+	return dt_from_rdn((int)(secs / SECS_PER_DAY) + DT_EPOCH_1970_OFFSET);
+}
+
+static int64_t
+local_secs(const struct datetime *date)
+{
+	return (int64_t)date->epoch + date->tzoffset * 60;
+}
+
+void
+datetime_to_tm(const struct datetime *date, struct datetime_tm *tm)
+{
+	struct tm t;
+	memset(&t, 0, sizeof(t));
+	tm->tm_epoch = local_secs(date);
+	dt_to_struct_tm(local_dt(tm->tm_epoch), &t);
+	tm->tm_year = t.tm_year;
+	tm->tm_mon = t.tm_mon;
+	tm->tm_mday = t.tm_mday;
+	tm->tm_wday = t.tm_wday;
+	tm->tm_yday = t.tm_yday;
+
+	tm->tm_gmtoff = date->tzoffset * 60;
+	tm->tm_nsec = date->nsec;
+
+	int seconds_of_day = tm->tm_epoch % SECS_PER_DAY;
+	tm->tm_hour = (seconds_of_day / 3600) % 24;
+	tm->tm_min = (seconds_of_day / 60) % 60;
+	tm->tm_sec = seconds_of_day % 60;
+}
+
+size_t
+tnt_strftime(char *s, size_t maxsize, const char *format,
+	     const struct datetime_tm *t)
+{
+	char *p;
+	int saved_errno = errno;
+	enum warn warn = IN_NONE;
+
+	tzset();
+	p = _fmt(format, t, s, s + maxsize, &warn);
+	if (!p) {
+		errno = EOVERFLOW;
+		return 0;
+	}
+	if (p == s + maxsize) {
+		errno = ERANGE;
```

Comment:
I added here `assert(false)` - the tests passed. Please, cover this with a test.

Reply:
> ```c/**
>  * NB! buf may be NULL, and we should handle it gracefully, returning
> ```
> Still need a test here.

[Putting this comment here, because it's easier to put here, not there. But this is still relevant]

Here is lower-level test case, which I've created for checking of `tnt_datetime_to_string` and `tnt_datetime_strftime` boundary conditions handling.

```diff
diff --git a/test/app-tap/datetime.test.lua b/test/app-tap/datetime.test.lua
index d062bcc07..475fd30ab 100755
--- a/test/app-tap/datetime.test.lua
+++ b/test/app-tap/datetime.test.lua
@@ -3,8 +3,9 @@
 local tap = require('tap')
 local test = tap.test('errno')
 local date = require('datetime')
+local ffi = require('ffi')
 
-test:plan(12)
+test:plan(13)
 
 local function assert_raises(test, error_msg, func, ...)
     local ok, err = pcall(func, ...)
@@ -230,6 +231,26 @@ test:test("Simple date creation by attributes - check failed", function(test)
                   end)
 end)
 
+test:test("Formatting limits", function(test)
+    test:plan(6)
+    local T = date.new()
+    local len = ffi.C.tnt_datetime_to_string(T, nil, 0)
+    test:is(len, 20, 'tostring() with NULL')
+    local buff = ffi.new('char[?]', len + 1)
+    len = ffi.C.tnt_datetime_to_string(T, buff, len + 1)
+    test:is(len, 20, 'tostring() with non-NULL')
+    test:is(ffi.string(buff), '1970-01-01T00:00:00Z', 'Epoch string')
+
+    local fmt = '%d/%m/%Y'
+    len = ffi.C.tnt_datetime_strftime(T, nil, 0, fmt)
+    test:is(len, 0, 'format(fmt) with NULL')
+    local strfmt_sz = 128
+    buff = ffi.new('char[?]', strfmt_sz)
+    len = ffi.C.tnt_datetime_strftime(T, buff, strfmt_sz, fmt)
+    test:is(len, 10, 'format(fmt) with non-NULL')
+    test:is(ffi.string(buff), '01/01/1970', 'Epoch string (fmt)')
+end)
+
 local function expected_str(msg, value)
     return ("%s: expected string, but received %s"):format(msg, value)
 end
```

Please pay attention that those 2 functions behave differently for the case when nil passed for buffer, and 0 for the length. `tnt_datetime_to_string` was adapted for `mp_snprint` (see future patches)  so it handles such case, and we have ability to calculate the length of target buffer first. But `tnt_datetime_strftime` is adaptation of `strftime` from LIBC, thus it behaves differently, and simply return 0 for NULL buffer. 

This inconsistency would not a bid deal, if we could handle huge buffers gracefully, but we could not. 

So here is current plan, as suggested by @Gerold103:
- return final size in each case, whether we have buffer or not;
- if continue calculation of a size even if currently passed buffer size is not enough (since some moment);
- in this case we still have optimal performance if we correctly guessed buffer size inside of `datetie_strftime`, and if size of buffer was not enough, we could reallocate and restart.

Reply:
Let's call it 1.5 passes of `strftime`